src/aio/tokio.rs

```diff
@@ -44,34 +44,72 @@ impl Provider for Tokio {
     }
 }
 
+/// Timeout for each broadcast response.
+pub const RESPONSE_TIMEOUT: Duration = Duration::from_secs(10);
+
 /// Search for a gateway with the provided options.
 pub async fn search_gateway(options: SearchOptions) -> Result<Gateway<Tokio>, SearchError> {
     // Create socket for future calls
     let mut socket = UdpSocket::bind(&options.bind_addr).await?;
 
     send_search_request(&mut socket, options.broadcast_address).await?;
 
-    let search_response = receive_search_response(&mut socket);
-
-    // Receive search response, optionally with a timeout.
-    let (response_body, from) = match options.timeout {
-        Some(t) => timeout(t, search_response).await?,
-        None => search_response.await,
-    }?;
-
-    let (addr, root_url) = handle_broadcast_resp(&from, &response_body)?;
-
-    let (control_schema_url, control_url) = get_control_urls(&addr, &root_url).await?;
-    let control_schema = get_control_schemas(&addr, &control_schema_url).await?;
-
-    Ok(Gateway {
-        addr,
-        root_url,
-        control_url,
-        control_schema_url,
-        control_schema,
-        provider: Tokio,
-    })
+    let max_search_time = options.timeout.unwrap_or(DEFAULT_TIMEOUT);
+    let start_search_time = std::time::Instant::now();
+
+    loop {
+        if start_search_time.elapsed() > max_search_time {
+            break Err(SearchError::NoResponseWithinTimeout);
+        }
```

Comment:
I wonder if it would be better to use `timeout` from tokio for this loop instead of the check. Thoughts?

Reply:
I think the code is more readable like this than when using tokio::timeout for the loop. 

Also, this way we do not cut-off in progress responses (even if the max search time has been exceeded). Worst case scenario, the search time would be the `max_search_time + RESPONSE_TIMEOUT`.

If you want the max search time to be exactly what you specify, then a tokio::timeout would be better.

Reply:
if we go this way, then instead of `loop` I suggest using `while start_search_time.elapsed() < max_search_time` replacing the `break`s with just `return`

Reply:
> if we go this way, then instead of `loop` I suggest using `while start_search_time.elapsed() < max_search_time` replacing the `break`s with just `return`

A `while loop` can't evaluate to a value. So the code would have to be changed to something like this:

```rust
let mut maybe_gateway: Option<Gateway<Tokio>> = None;

while start_search_time.elapsed() < max_search_time && maybe_gateway.is_none() {
    // ...
    maybe_gateway = Some(Gateway {
        addr,
        root_url,
        control_url,
        control_schema_url,
        control_schema,
        provider: Tokio,
    });
}

if let Some(gateway) = maybe_gateway {
    Ok(gateway)
} else {
    Err(SearchError::NoResponseWithinTimeout)
}
```

Which is maybe less clean than the original `loop` solution?

Reply:
Sorry for the late response.

> I think the code is more readable like this than when using tokio::timeout for the loop.
> 
> Also, this way we do not cut-off in progress responses (even if the max search time has been exceeded). Worst case scenario, the search time would be the `max_search_time + RESPONSE_TIMEOUT`.
> 
> If you want the max search time to be exactly what you specify, then a tokio::timeout would be better.

Looking back, we could probably stick with the loop so what we have in this PR is fine :).

> A `while loop` can't evaluate to a value. So the code would have to be changed to something like this:
> 
> ```rust
> let mut maybe_gateway: Option<Gateway<Tokio>> = None;
> 
> while start_search_time.elapsed() < max_search_time && maybe_gateway.is_none() {
>     // ...
>     maybe_gateway = Some();
> }
> 
> if let Some(gateway) = maybe_gateway {
>     Ok(gateway)
> } else {
>     Err(SearchError::NoResponseWithinTimeout)
> }
> ```
> 
> Which is maybe less clean than the original `loop` solution?

I believe he means something like 

```rust
while start_search_time.elapsed() < max_search_time {
    // ....
    return Ok(Gateway {
         addr,
         root_url,
         control_url,
         control_schema_url,
         control_schema,
         provider: Tokio,
     });
}

Err(SearchError::NoResponseWithinTimeout)

```

Reply:
Ah thank you, I was somehow under the misconception that you can not return a value from within a while loop 😅.

Made the changes!